### ISSUE TITLE
pip-sync keeps collecting and installing different versions of `decorator`

### ISSUE THREAD

**Issue by chaoflow** [nvie/pip-tools#318]
pip-tools-1.6, pip-8.0.2, setuptools-20.0

An activated virtualenv
```
% virtualenv .
% . ./bin/activate
```

Pinned requirements for ipdb, most notably `decorator==4.0.8`
```
% cat req.txt
#
# This file is autogenerated by pip-compile
# Make changes in req.in, then run this to update:
#
#    pip-compile req.in
#
decorator==4.0.8          # via ipython, traitlets
ipdb==0.8.1
ipython-genutils==0.1.0   # via traitlets
ipython==4.1.1            # via ipdb
path.py==8.1.2            # via pickleshare
pexpect==4.0.1            # via ipython
pickleshare==0.6          # via ipython
ptyprocess==0.5.1         # via pexpect
simplegeneric==0.8.1      # via ipython
traitlets==4.1.0          # via ipython

# The following packages are commented out because they are
# considered to be unsafe in a requirements file:
# setuptools                # via ipython
```

`decorator==4.0.8` is collected, but `decorator==4.0.7` is installed.
```
% pip-sync req.txt 
Collecting decorator==4.0.8
  Using cached decorator-4.0.8-py2.py3-none-any.whl
Collecting ipdb==0.8.1
  Using cached ipdb-0.8.1.zip
Collecting ipython-genutils==0.1.0
  Using cached ipython_genutils-0.1.0-py2.py3-none-any.whl
Collecting ipython==4.1.1
  Using cached ipython-4.1.1-py2-none-any.whl
Collecting path.py==8.1.2
  Using cached path.py-8.1.2-py2.py3-none-any.whl
Collecting pexpect==4.0.1
  Using cached pexpect-4.0.1.tar.gz
Collecting pickleshare==0.6
  Using cached pickleshare-0.6-py2.py3-none-any.whl
Collecting ptyprocess==0.5.1
  Using cached ptyprocess-0.5.1-py2.py3-none-any.whl
Collecting simplegeneric==0.8.1
  Using cached simplegeneric-0.8.1.zip
Collecting traitlets==4.1.0
  Using cached traitlets-4.1.0-py2.py3-none-any.whl
Requirement already satisfied (use --upgrade to upgrade): setuptools>=18.5 in ./lib/python2.7/site-packages (from ipython==4.1.1)
Installing collected packages: decorator, ipython-genutils, traitlets, path.py, pickleshare, simplegeneric, ptyprocess, pexpect, ipython, ipdb
  Running setup.py install for simplegeneric ... done
  Running setup.py install for pexpect ... done
  Running setup.py install for ipdb ... done
Successfully installed decorator-4.0.7 ipdb-0.8.1 ipython-4.1.1 ipython-genutils-0.1.0 path.py-8.1.2 pexpect-4.0.1 pickleshare-0.6 ptyprocess-0.5.1 simplegeneric-0.8.1 traitlets-4.1.0
```

pip-sync keeps collecting and installing different `decorator` versions.
```
% pip-sync req.txt
Collecting decorator==4.0.8
  Using cached decorator-4.0.8-py2.py3-none-any.whl
Installing collected packages: decorator
  Found existing installation: decorator 4.0.7
    Uninstalling decorator-4.0.7:
      Successfully uninstalled decorator-4.0.7
Successfully installed decorator-4.0.7
% pip-sync req.txt
Collecting decorator==4.0.8
  Using cached decorator-4.0.8-py2.py3-none-any.whl
Installing collected packages: decorator
  Found existing installation: decorator 4.0.7
    Uninstalling decorator-4.0.7:
      Successfully uninstalled decorator-4.0.7
Successfully installed decorator-4.0.7
```

**Comment by nvie** [comment]
Sounds like a bug in `pip` to me:

```console
$ pip install decorator==4.0.8
Collecting decorator==4.0.8
  Using cached decorator-4.0.8-py2.py3-none-any.whl
Installing collected packages: decorator
  Found existing installation: decorator 4.0.7
    Uninstalling decorator-4.0.7:
      Successfully uninstalled decorator-4.0.7
Successfully installed decorator-4.0.7
```

This happens even without the use of pip-tools. Perhaps report over at pypa/pip?

**Comment by chaoflow** [comment]
It seems this was an issue with decorator-4.0.8, fixed with decorator-4.0.9. Sorry for the noise.

**Comment by nvie** [comment]
No problem :+1:
